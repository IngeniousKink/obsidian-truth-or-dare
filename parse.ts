import type { Root as MarkdownRoot } from "mdast-util-from-markdown/lib";

// Remember to rename these classes and interfaces!

export function getCardsUnderHeading(node: MarkdownRoot) {
	let headingText = '';

	const allCardsUnderHeadings = {};

	node.children.forEach(child => {
		let refCounter = 0;
		const cardsUnderHeading: { ref: string; text: any; }[] = [];

		if (child.type === 'heading' && child.children) {
			headingText = child.children[0].value;
			return;
		}

		if (child.type !== 'list') {
			return;
		}

		child.children.forEach(listItem => {
			listItem.children.forEach(paragraph => {

				paragraph.children.forEach(textNode => {
					if (textNode.type === 'text') {

						const filePath = '';
						cardsUnderHeading.push({
							ref: filePath + '#' + headingText + '^' + refCounter++,
							text: textNode.value,
						});
					}
				});
			});
		});

		console.log(`Cards under heading "${headingText}":`, cardsUnderHeading);

		allCardsUnderHeadings[headingText] = cardsUnderHeading;

	});

	return allCardsUnderHeadings;
}
