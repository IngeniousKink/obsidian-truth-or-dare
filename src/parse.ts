import type { Root, Heading, List, ListItem, Paragraph, Text } from "mdast";

// Define a Card type with a reference and text
export type Card = { ref: string; text: string; };

// Define a map to hold all cards indexed by their reference
export type CardMap = Record<string, Card[]>;

// ChildNode type can be either a Heading or a List
type ChildNode = Heading | List;

/**
 * This function takes a markdown abstract syntax tree (AST) and converts it into a game state.
 * The game state is represented as a CardMap, where each card is indexed by a unique reference.
 * Each heading in the markdown creates a new key in the CardMap, and each list item under that heading
 * creates a new card under that key.
 * 
 * For example, given the following markdown:
 * 
 * # Heading 1
 * - Item 1
 * - Item 2
 * 
 * # Heading 2
 * - Item 3
 * 
 * The resulting CardMap would be:
 * 
 * {
 *   "Heading 1": [
 *     { ref: "#Heading 1^0", text: "Item 1" },
 *     { ref: "#Heading 1^1", text: "Item 2" }
 *   ],
 *   "Heading 2": [
 *     { ref: "#Heading 2^0", text: "Item 3" }
 *   ]
 * }
 */
export function convertMarkdownToGameTemplate(node: Root): CardMap {
  return node.children.reduce<CardMap>(
    (cardMap, child) => processMarkdownNode(cardMap, child as ChildNode),
    {}
  );
}

// Process a markdown node, which can either be a heading or a list
function processMarkdownNode(cardMap: CardMap, child: ChildNode): CardMap {
  if (child.type === 'heading') {
    const headingText = extractHeadingText(child);
    return { ...cardMap, [headingText]: [] };
  }

  if (child.type === 'list') {
    const headingText = getLastHeading(cardMap);
    const cardsUnderHeading = extractCardsFromList(child, headingText);
    return { ...cardMap, [headingText]: cardsUnderHeading };
  }

  return cardMap;
}

// Extract the text from a heading node
function extractHeadingText(node: Heading): string {
  const textNode = node.children[0] as Text;
  return textNode.value || '';
}

// Get the last heading from the card map
function getLastHeading(cardMap: CardMap): string {
  return Object.keys(cardMap).pop() || '';
}

// Extract cards from a list node
function extractCardsFromList(listNode: List, headingText: string): Card[] {
  let refCounter = 0;
  return listNode.children.reduce<Card[]>((cards, listItem: ListItem) => {
    const listItemCards = extractCardsFromListItem(listItem, refCounter, headingText);
    return [...cards, ...listItemCards];
  }, []);
}

// Extract cards from a list item
function extractCardsFromListItem(listItem: ListItem, refCounter: number, headingText: string): Card[] {
  return listItem.children.reduce<Card[]>((cards, paragraph: Paragraph) => {
    const paragraphCards = extractCardsFromParagraph(paragraph, refCounter, headingText);
    return [...cards, ...paragraphCards];
  }, []);
}

function extractCardsFromParagraph(paragraph: Paragraph, refCounter: number, headingText: string): Card[] {
  return paragraph.children.reduce<Card[]>((cards, textNode: Text) => {
    if (textNode.type !== 'text') return cards;

    const card = {
      ref: '#' + headingText + '^' + refCounter++,
      text: textNode.value,
    };
    return [...cards, card];
  }, []);
}