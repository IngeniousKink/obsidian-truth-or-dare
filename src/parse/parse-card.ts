import { PhrasingContent } from "node_modules/mdast-util-from-markdown/lib/index.js";

export interface Annotation {
  actor?: string;
  startPos?: number;
  endPos?: number;
  [key: string]: string | number | undefined;
}

export interface ParsedCard {
  text: string;
  annotations: Annotation[];
}

const REGEX_HTML_DATA_ATTRIBUTE = /<([^>]+?)>/g;
const REGEX_DATA_ATTRIBUTE = /data-(\w+)="([^"]*)"/g;
const REGEX_MEDIA_ATTRIBUTE = /<(img|video)\s+src="([^"]*)"/g;
const REGEX_OBSIDIAN_EMBED = /!\[\[([^]|]+)(?:\|[^]]+)?\]\]/g;

function createAnnotation(
  dataMatches: RegExpMatchArray[],
  match: RegExpMatchArray,
  annotationData?: Partial<Annotation>
): Annotation {
  const annotation: Partial<Annotation> = annotationData || {};
  dataMatches.forEach((dataMatch) => {
    annotation[dataMatch[1]] = dataMatch[2];
  });
  annotation.startPos = match.index;
  annotation.endPos = ((match.index || 0) + match[0].length);
  return annotation;
}

function parseDataAttributes(match: RegExpMatchArray): Annotation | null {
  const dataMatches = Array.from(match[1].matchAll(REGEX_DATA_ATTRIBUTE));
  if (dataMatches.length === 0) return null;
  return createAnnotation(dataMatches, match);
}

function parseMedia(match: RegExpMatchArray): Annotation {
  const annotationData: Partial<Annotation> = {
    src: match[2],
    type: match[1],
  };
  return createAnnotation([], match, annotationData);
}

function parseObsidianEmbed(match: RegExpMatchArray): Annotation {
  const annotationData: Partial<Annotation> = {
    src: match[1],
    type: 'obsidian',
  };
  return createAnnotation([], match, annotationData);
}

export function parseCard(content: PhrasingContent[]): ParsedCard {
  let position = 0;
  const annotations: Annotation[] = [];
  
  const html = content.reduce((acc, textNode: PhrasingContent) => {
    let value = '';
    if (textNode.type === 'image' && textNode.position?.start.column) {
      const annotation: Annotation = {
        type:'image',
        'url': textNode.url,
        startPos: position,
        endPos: position + '![['.length + textNode.url.length + ']]'.length,
      };
      annotations.push(annotation);
      value = '![[' + textNode.url + ']]';
    } else if (textNode.type === 'text' || textNode.type === 'html') {
      value = textNode.value;
      const dataMatches = Array.from(value.matchAll(REGEX_HTML_DATA_ATTRIBUTE));
      dataMatches.forEach(match => {
        const annotation = parseDataAttributes(match);
        if (annotation) {
          annotation.startPos = (annotation.startPos || 0) + position;
          annotation.endPos = (annotation.endPos || 0) + position;
          annotations.push(annotation);
        }
      });
      const mediaMatches = Array.from(value.matchAll(REGEX_MEDIA_ATTRIBUTE));
      mediaMatches.forEach(match => {
        const annotation = parseMedia(match);
        annotation.startPos = (annotation.startPos || 0) + position;
        annotation.endPos = (annotation.endPos || 0) + position;
        annotations.push(annotation);
      });
      const obsidianMatches = Array.from(value.matchAll(REGEX_OBSIDIAN_EMBED));
      obsidianMatches.forEach(match => {
        const annotation = parseObsidianEmbed(match);
        annotation.startPos = (annotation.startPos || 0) + position;
        annotation.endPos = (annotation.endPos || 0) + position;
        annotations.push(annotation);
      });
    }
    position += value.length;
    return acc + value;
  }, '');
  
  return {
    text: html,
    annotations: annotations,
  };
}
