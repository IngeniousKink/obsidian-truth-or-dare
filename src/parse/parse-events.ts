import { AdvanceStackEvent, AssignSlotEvent, ChangeActorInventoryAddEvent, ChangeActorInventoryRemoveEvent, ChangeActorNameEvent, CompleteCardEvent, CreateActorEvent, DeleteActorEvent, DrawCardEvent, GameEvent, SkipCardEvent, TimestampedEvent, UnassignSlotEvent, WithTimestamp } from "@obsidian-truth-or-dare/events.js";
import type { Root, Code } from "mdast";

/**
 * This function takes a markdown abstract syntax tree (AST) and converts it into an array of game events.
 * The game events are represented as a GameEvent, where each event is an object with a game, type, timestamp and x.
 * Each code block in the markdown creates a new event in the GameEvent array.
 * 
 * For example, given the following markdown:
 * 
 * ```truth-or-dare:event
 * type:name-of-event
 * timestamp: 1706012788684
 * x: 2
 * ```
 * 
 * The resulting GameEvent would be:
 * 
 * {
 *   game: "truth-or-dare",
 *   type: "name-of-event",
 *   timestamp: 1706012788684,
 *   x: 2
 * }
 */
export function convertMarkdownToGameEvents(node: Root): TimestampedEvent[] {
  return node.children.reduce<TimestampedEvent[]>(
    (gameEvents, child) => processMarkdownNode(gameEvents, child as Code),
    []
  );
}

// Process a markdown node, which can be a code block
function processMarkdownNode(gameEvents: TimestampedEvent[], child: Code): TimestampedEvent[] {
  if (child.type === 'code' && child.lang === 'truth-or-dare:event') {
    try {
      const event = extractEventFromCode(child as Code);
      return [...gameEvents, event];
    } catch(error) {
      if (error instanceof UnparseableEventError) {
        console.warn(error);
      } else {
        throw error;
      }
    }
  }

  return gameEvents;
}

export class UnparseableEventError extends Error {
  constructor(public event: any) {
    super(`Unable to parse the event ${event.type ? `of type ${event.type}` : ''} ${JSON.stringify(event)}`);
    this.name = "UnparseableEventError";
    this.event = event;
  }
}

// Extract the game event from a code block
function extractEventFromCode(node: Code): TimestampedEvent {
  const lines = node.value.split('\n');
  const event: any = {};

  for (const line of lines) {
    const [key, value] = line.split(':');

    if (key && value) {
      event[key.trim()] = value.trim();
    }
  }

  if (!('timestamp' in event)) {
      throw new UnparseableEventError(event);
  }

  switch (event.type) {
    case 'create_actor':
      if ('actorId' in event) {
        return event as CreateActorEvent & WithTimestamp;
      }
      break;
    case 'delete_actor':
      if ('actorId' in event) {
        return event as DeleteActorEvent & WithTimestamp;
      }
      break;
    case 'change_actor_name':
      if ('actorId' in event && 'value' in event) {
        return event as ChangeActorNameEvent & WithTimestamp;
      }
      break;
    case 'change_actor_inventory_add':
      if ('actorId' in event && 'value' in event) {
        return event as ChangeActorInventoryAddEvent & WithTimestamp;
      }
      break;
    case 'change_actor_inventory_remove':
      if ('actorId' in event && 'value' in event) {
        return event as ChangeActorInventoryRemoveEvent & WithTimestamp;
      }
      break;
    case 'draw_card':
      if ('cardRef' in event) {
        return event as DrawCardEvent & WithTimestamp;
      }
      break;
    case 'skip_card':
      if ('cardId' in event) {
        return event as SkipCardEvent & WithTimestamp;
      }
      break;
    case 'complete_card':
      if ('cardId' in event) {
        return event as CompleteCardEvent & WithTimestamp;
      }
      break;
    case 'assign_slot':
      if ('actorId' in event && 'slotIndex' in event) {
        return event as AssignSlotEvent & WithTimestamp;
      }
      break;
    case 'unassign_slot':
      if ('slotIndex' in event) {
        return event as UnassignSlotEvent & WithTimestamp;
      }
      break;
    case 'advance_stack':
      return event as AdvanceStackEvent & WithTimestamp;
    default:
      throw new UnparseableEventError(event);
  }

  throw new UnparseableEventError(event);
}

