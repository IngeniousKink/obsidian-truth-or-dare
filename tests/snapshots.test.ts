
import { fromMarkdown } from 'mdast-util-from-markdown';
import { getCardsUnderHeading } from '../parse.js';

describe('getCardsUnderHeading', () => {
    it('renders correctly', () => {
        const fileContents = `
# Heading
* card 1
* card 2
        `;
        const mast = fromMarkdown(fileContents);
        const result = getCardsUnderHeading(mast);

        // Match the result to a snapshot
        expect(result).toMatchSnapshot();
    });
});
