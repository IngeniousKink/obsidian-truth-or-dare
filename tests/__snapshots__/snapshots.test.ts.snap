// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCardsUnderHeading renders correctly 1`] = `
{
  "": [
    {
      "ref": "#^0",
      "text": "card 1",
    },
    {
      "ref": "#^1",
      "text": "card 2",
    },
  ],
}
`;
